### ISSUE TITLE
is there a way to run xonsh from .py script as .xsh script?

### ISSUE THREAD

**Issue by andry81** [xonsh/xonsh#2478]
I want to pack xonsh instructions under the `.py` extension to run xonsh as usual python script:
`python myscript.py`

There `myscript.py` is something like this:
```
if __name__ == '__main__':
	if not 'builtins' in globals() or not hasattr(builtins, '__xonsh_env__'):
		from xonsh import main as xonsh
		# some magic to force treat `myscript.py` as `myscript.xsh`
		xonsh.main()

if 'builtins' in globals() and hasattr(builtins, '__xonsh_env__'):
	print('hello xonsh\n')
	dir /A:D /B
	1 + 1
```

Is there a way to do so?

**Comment by astronouth7303** [comment]
Xonsh scripts must be compiled as xonsh in order to run.

However, importing `xonsh.imphooks` makes `.xsh` files importable.

**Comment by andry81** [comment]
I've found closest solution:
```
if __name__ == '__main__':
	import builtins
	if not 'builtins' in globals() or not hasattr(builtins, '__xonsh_shell__'):
		from xonsh import main as xonsh
		import sys
		sys.argv.insert(0, __file__)
		print('main')
		xonsh.main(sys.argv)
		sys.exit()

#import builtins
if 'builtins' in globals() and hasattr(builtins, '__xonsh_shell__'):
	print('in xonsh\n')
	evalx('dir /A:D /B "*.*"')
	print(evalx('1 + 1'))
```

I works for both:
`python test.py`
`python -m xonsh test.py`

Is there a better way without the `evalx`? Or may be through the `@` without quotes?

**Comment by astronouth7303** [comment]
Write a trampoline script that imports `xonsh.imphooks` and then calls to an `.xsh` file.

**Comment by andry81** [comment]
It's a bit strange to use `.xsh` through the `.py`. Why to do that if you can directly call to `.xsh`? It has no sense.

**Comment by astronouth7303** [comment]
If you want to execute xonsh scripts directly, you have to use xonsh.

If you want to call xonsh code from python, you must first bootstrap the xonsh parser.

**Comment by andry81** [comment]
I don't get the difference. Is that call is a call `from python` or `using xonsh`?
`python -m xonsh test.xsh`

My point was to use single file of `.py` extension for all things including xonsh code. Using 2 extensions has no sense because i can call `.xsh` directly then.

**Comment by astronouth7303** [comment]
`python -m xonsh` works for xonsh, if the python environment has xonsh installed.

Xonsh can perfectly understand xonsh and python files.

**Comment by andry81** [comment]
Then why i need to bootstrap anything?

**Comment by astronouth7303** [comment]
Your original question was to get `python myscript.py` working.

**Comment by andry81** [comment]
Yes. But then i have to move a half of logic to another file with `.xsh` extension instead of use the `.xsh` file as a single file.

**Comment by astronouth7303** [comment]
Sorry, we don't have a magic way to recompile the file in the middle of its own execution.

**Comment by andry81** [comment]
But why not? Is there a function to do so? Or may be i am asking too much?

**Comment by astronouth7303** [comment]
1. Use the subset of xonsh that's still parsable as Python (throwing away command execution and macros)
2. Python compiles the file to bytecode and begins executing
3. The module calls a function, which reparses and recompiles the file as xonsh, somehow manages to move the current instruction to the new bytecode, while transitioning all of the state. Without any bugs.
4. Execution continues as if nothing happened.

In summary: That's a research thesis, and we're a small team of volunteers.

**Comment by andry81** [comment]
I thought it could be solved somehow via on runtime self compilation to different bytecode w/o any injection to existing one and start execution a new one from beginning. But of cause it has sense only when the runtime runs from a script.
Anyway, thx for the clarification!

**Comment by astronouth7303** [comment]
If you know enough about Python interpreter internals to make it happen, pull requests are always welcome.

**Comment by gforsyth** [comment]
Closing this as I believe the question is answered, but please re-open if there are further points/questions!

**Comment by v217** [comment]
There is no editor/ide support for xonsh xsh files. So is it possible to import some of the conveniences of xonsh into python, what I mean is eg:

    from xonsh import xonsh_subprocess_run

So you don't get the convenient syntax of xonsh but just a few of it's convenient functions etc.
I don't want to learn about a second package like eg:
http://amoffat.github.io/sh/
or
plumbum
just for scripts.
I couldn't find docs on this. I guess for python experts the way to do this is obvious.

**Comment by scopatz** [comment]
This is somewhat undocumented, but if you want to start up a xonsh environment, for 3rd party packages, I often put the following in the `__init__.py` file, at the very top:

```python 
from xonsh.main import setup
setup()
del setup
```

This is safely reentrant and will set up the xonsh session object for you, if it hasn't been already.  Here is an example in the wild: https://github.com/regro/rever/blob/master/rever/__init__.py

**Comment by v217** [comment]
Thank you, this works!

**Comment by scopatz** [comment]
Awesome! We should really add this to the docs....

**Comment by v217** [comment]
Here my documentation effort:
I am not sure about the best place in the docs - maybe just a link to this issue.  
I have summarized your answer, so a python novice unfamiliar with python module conventions understands how to use xonsh modules in python.

## Can I call xonsh from my python project?

Sure, suppose you have some xonsh files `file1.xsh`, `file2.xsh` and the file `__init__.py` with the following content:

```python
# __init__.py
from xonsh.main import setup
setup()
del setup
```

in the directory `my_xonsh_directory` which must be available in your `PYTHON_PATH` eg. in the same directory like your `python_file.py`:

```
./python_file.py

./my_xonsh_directory

./my_xonsh_directory/__init__.py
./my_xonsh_directory/file1.xsh
./my_xonsh_directory/file2.xsh
``` 
then you can import eg. `file1.xsh` in `python_file.py` exactly like you would import a normal python file:


```python
# ./python_file.py
from my_xonsh_directory import file1

```

**Comment by scopatz** [comment]
Thanks, I have already opened #3332 for more docs here
